### PR TITLE
Unseal blocks streamed from peers and track processed channel items without spawning a goroutine

### PR DESCRIPTION
**Pull request by robskillington**
cc @xichen2020 @martin-mao

### CODE REVIEW

File: client/session.go [hunk 1]
```diff
@@ -1252,6 +1263,7 @@ func (r *blocksResult) addBlockFromPeer(id string, block *rpc.Block) error {
 		defer iter.Close()
 
 		encoder := r.encoderPool.Get()
+		encoder.Reset(start, r.blockAllocSize)
 		defer encoder.Close()
```
Comment (xichen2020) [inline client/session.go]: Are we sure we want to close the encoder here?

File: client/session.go [hunk 2]
```diff
@@ -1492,6 +1502,15 @@ func (b blocksMetadata) unselectedBlocks() []blockMetadata {
 type blocksMetadatas []*blocksMetadata
 
 func (arr blocksMetadatas) swap(i, j int) { arr[i], arr[j] = arr[j], arr[i] }
+func (arr blocksMetadatas) hasBlocksLen() int {
+	count := 0
+	for i := range arr {
+		if len(arr[i].blocks) > 0 {
```
Comment (xichen2020) [inline client/session.go]: Can `arr[i]` ever be nil?

File: client/session.go (hunk 1 above)
Comment (robskillington) [inline client/session.go]: Yeah probably not lol, will update and also assert not closed by pulling the test encoder out and checking it's state.

File: client/session.go (hunk 2 above)
Comment (robskillington) [inline client/session.go]: Should not be, but I'll add this check anyway.

File: client/session.go (hunk 1 above)
Comment (xichen2020) [inline client/session.go]: Sounds good

**Comment by xichen2020** [comment]
LGTM

**Pull request merged**